nupic/research/frameworks/pytorch/datasets/permuted_mnist.py

```diff
@@ -0,0 +1,145 @@
+        super().__init__(root=root, train=train, transform=data_transform,
+                         target_transform=target_transform, download=download)
+
+        self.num_tasks = num_tasks
+
+        # Generate num_tasks random permutations - the first one is the identity
+        # permutation (i.e., regular MNIST), represented below as `None`
+        if PermutedMNIST.permutations is None:
+            PermutedMNIST.permutations = [
+                torch.randperm(784) for task_id in range(1, num_tasks)
+            ]
+            PermutedMNIST.permutations.insert(0, None)
+
+        self.permutations = PermutedMNIST.permutations
+
+    def __getitem__(self, index):
+        """
+        Returns an (image, target) pair.
+
+        In particular, this method retrieves an MNIST image, and based on the value of
+        `index`, determines which pixel-wise permutation to apply. Target values are
+        also scaled to be unique to each permutation.
+        """
+        img, target = super().__getitem__(index % len(self.data))
+
+        # Determine which task `index` corresponds to
+        task_id = index // len(self.data)
+
+        # Apply permutation to `img`
+        img = permute(img, self.permutations[task_id])
+
+        # Since target values are not shared between tasks, `target` should be in the
+        # range [0 + 10 * task_id, 9 + 10 * task_id]
+        target += 10 * task_id
+
+        # Since image size is not relevant for an MLP, `img` can be flattened
+        return img.flatten(), target
+
+    def __len__(self):
+        return self.num_tasks * len(self.data)
+
+    @property
+    def processed_folder(self):
+        return os.path.join(self.root, "MNIST", "processed")
+
+
+class ContextDependentPermutedMNIST(PermutedMNIST):
+    """
+    A variant of permutedMNIST where each permutation (i.e., 'task') is associated with
+    a context: a binary sparse vector. The `__getitem__` method returns the context
+    vector along with the data sample and target.
+    """
+
+    # Context vectors should be the same whether this class is instantiated for
+    # training or testing
+    contexts = None
+
+    def __init__(self, num_tasks, dim_context, train, root=".", target_transform=None,
+                 download=False):
+
+        super().__init__(num_tasks, train, root, target_transform, download)
+
+        # Generate context vectors if `contexts` isn't already defined
+        if ContextDependentPermutedMNIST.contexts is None:
+            ContextDependentPermutedMNIST.contexts = generate_context_vectors(
+                num_contexts=num_tasks, n_dim=dim_context, percent_on=0.05
+            )
+
+        self.contexts = ContextDependentPermutedMNIST.contexts
+
+    def __getitem__(self, index):
+        """
+        Returns an (image, context, target) triplet.
+        """
+        img, target = super().__getitem__(index)
+        task_id = index // len(self.data)
+        return img, self.contexts[task_id, :], target
```

Comment:
Maybe also return `task_id` here?

Reply:
What's the need? If the context vector is provided, then that itself is task information. If we return `task_id`, then it should be eventually used to select the appropriate context vector, which means contexts would be separate from the dataset class. That would make sense, but it still seems better to do it this way.

Reply:
Ok, makes sense.

Reply:
I don't think modifying the signature of dataloader getitem is a good alternative. Dataloader is just data and target, and this case your data happens to be not just an image, but an image but a context. It could potentially be a composition of even more things.

Python is not type enforced and type hints are commonly used with unions of types, that flexibility exists by design. So data and target could be of different types. I think the best alternative here is returning a tuple data (img, context) and target, preserving the original signature.

Reply:
I don't think modifying the signature of dataloader getitem is a good alternative. Dataloader is just data and target, and this case your data happens to be not just an image, but an image but a context. It could potentially be a composition of even more things.

Python is not type enforced and type hints are commonly used with unions of types, that flexibility exists by design. So data and target could be of different types. I think the best alternative here is returning a tuple data (img, context) and target, preserving the original signature.

Reply:
@karangrewal just making sure you saw this comment as well

Reply:
@lucasosouza noted! I can make this change so that the dataloader yields ((data, context), target) tuples.

Reply:
Fixed.